### PR TITLE
[1.19] Add Ability for Recipes to define Item Remainders only for that Recipe

### PR DESCRIPTION
**Pull request by Dragon-Seeker**
Bonus is a Deep Copy Method for Item.Settings

### CODE REVIEW

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @Inject(method = "deserialize", at = @At(value = "RETURN"))
+    private static void deserializeRecipeSpecificRemainders(Identifier id, JsonObject json, CallbackInfoReturnable<Recipe<?>> cir){
+        if(json.has("owo:remainders")){
+            JsonObject remainders = json.getAsJsonObject("owo:remainders");
+
+            for(Map.Entry<String, JsonElement> entry : remainders.entrySet()){
+                Identifier item_id = Identifier.tryParse(entry.getKey());
+
+                if(item_id == null) {
+                    throw new JsonSyntaxException("OwO Remainders: A recipe Remainder has invalid item Identifier");
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: It's spelled `oωo`

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                }
+
+                Optional<Item> item = Registry.ITEM.getOrEmpty(item_id);
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: `JsonHelper.asItem()`

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeMixin.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Set;
+
+@Mixin(Recipe.class)
+public interface RecipeMixin<C extends Inventory> {
+
+    @Shadow Identifier getId();
+
+    @Inject(method = "getRemainder", at = @At("RETURN"), locals = LocalCapture.CAPTURE_FAILHARD)
+    private void addRecipeSpecificRemainders(C inventory, CallbackInfoReturnable<DefaultedList<ItemStack>> cir, DefaultedList<ItemStack> defaultedList) {
+        if (RecipeSpecificRemainderStorage.hasRecipeRemainders(getId())) {
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/mixin/recipe/RecipeMixin.java]: Please turn this into a return guard to reduce nesting

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeMixin.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            Set<RecipeSpecificRemainderStorage.RecipeRemainder> remainders = RecipeSpecificRemainderStorage.getRemainders(getId());
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/mixin/recipe/RecipeMixin.java]: `var` time

File: src/main/java/io/wispforest/owo/util/RecipeSpecificRemainderStorage.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    private static final Map<Identifier, Set<RecipeRemainder>> allSpecificRemainders = new HashMap<>();
+
+    /**
+     *  Method used to register a Recipe specific Item remainder for a given a Recipe
+     *
+     * @param recipeID The Recipe's Identifier
+     * @param item The Item to check for
+     * @param itemRemainder The Remainder Item for the given Item
+     */
+    public static void registerRecipeSpecificRemainder(Identifier recipeID, Item item, Item itemRemainder){
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/util/RecipeSpecificRemainderStorage.java]: The convention for the ID shorthand is `Id`, as such:  `recipeID` -> `recipeId`

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+                if(item.isEmpty()) {
+                    throw new JsonSyntaxException("OwO Remainders: The given Item could not be found within the Item Registry");
+                }
+
+                if (entry.getValue() instanceof JsonObject jsonObject) {
+                    if (jsonObject.has("item")) {
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: This `if` should ideally be inverted and throw immediately if `item` is not present - avoiding one level of nesting

File: src/main/java/io/wispforest/owo/util/RecipeSpecificRemainderStorage.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import io.wispforest.owo.mixin.recipe.RecipeMixin;
+import net.minecraft.item.Item;
+import net.minecraft.util.Identifier;
+
+import java.util.*;
+
+/**
+ *  Simple Class used to store and register all Recipe Specific Remainders.
+ */
+public final class RecipeSpecificRemainderStorage {
```
Comment (gliscowo) [inline src/main/java/io/wispforest/owo/util/RecipeSpecificRemainderStorage.java]: This class should be renamed to `RecipeSpecificRemainders`. Additonally, for consistency, all methods in it which specify further than `Remainer` should be shortened: `registerRecipeSpecificRemainder` -> `add`, `hasRecipeRemainders` -> `hasRemainders`

**Review by gliscowo — CHANGES_REQUESTED** [review]
Implementation looks fine, yet once again there are some general codestyle problems that need to be addressed before I will give approval

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java (hunk 2 above)
Comment (Dragon-Seeker) [reply src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: Not possible with how the API makes remainders: 
```
"owo:remainders": {
    "minecraft:sand": {
      "item": "minecraft:sand"
    }
  },
```

**Comment by Dragon-Seeker** [comment]
Would it be a good idea to have default behavior that if a empty JsonObject was given for the remainder item, it would default to use the given item as the Remainder?

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java (hunk 2 above)
Comment (Dragon-Seeker) [reply src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: Just slapped it into a JsonPrimitive and ran it into such

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java (hunk 1 above)
Comment (Dragon-Seeker) [reply src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: OwO

File: src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java (hunk 6 above)
Comment (Dragon-Seeker) [reply src/main/java/io/wispforest/owo/mixin/recipe/RecipeManagerMixin.java]: Fixed with using JsonHelper.getItem()

**Comment by Dragon-Seeker** [comment]
Could be such for specifying between the two: 

```
"owo:remainders": {
    "minecraft:sand": {
      "item": "minecraft:sand"
    },
   ...
   "item": "minecraft:sand"
  },
```

**Comment by Noaaan** [comment]
Might keep an eye on this PR to Fabric API, as it might supersede this: https://github.com/FabricMC/fabric/pull/2464

**Pull request closed without merging**